llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1531,22 +1531,31 @@ PassBuilder::buildPerModuleDefaultPipeline(OptimizationLevel Level,
 }
 
 ModulePassManager
-PassBuilder::buildFatLTODefaultPipeline(OptimizationLevel Level) {
+PassBuilder::buildFatLTODefaultPipeline(OptimizationLevel Level, bool ThinLTO,
+                                        bool EmitSummary) {
   ModulePassManager MPM;
-  // FatLTO always uses UnifiedLTO, so use the ThinLTOPreLink pipeline
-  MPM.addPass(buildThinLTOPreLinkDefaultPipeline(Level));
-  MPM.addPass(EmbedBitcodePass());
-
-  // Use the ThinLTO post-link pipeline with sample profiling, other
-  if (PGOOpt && PGOOpt->Action == PGOOptions::SampleUse)
-    MPM.addPass(buildThinLTODefaultPipeline(Level, /*ImportSummary=*/nullptr));
-  else {
-    // otherwise, just use module optimization
-    MPM.addPass(
-        buildModuleOptimizationPipeline(Level, ThinOrFullLTOPhase::None));
-    // Emit annotation remarks.
-    addAnnotationRemarksPass(MPM);
+  if (ThinLTO)
+    MPM.addPass(buildThinLTOPreLinkDefaultPipeline(Level));
+  else
+    MPM.addPass(buildLTOPreLinkDefaultPipeline(Level));
+  MPM.addPass(EmbedBitcodePass(ThinLTO, EmitSummary));
+
+  if (PGOOpt && PGOOpt->Action == PGOOptions::SampleUse) {
+    if (ThinLTO) {
+      // Use the ThinLTO post-link pipeline with sample profiling
+      MPM.addPass(
+          buildThinLTODefaultPipeline(Level, /*ImportSummary=*/nullptr));
+      return MPM;
+    }
+    MPM.addPass(SampleProfileLoaderPass(PGOOpt->ProfileFile,
+                                        PGOOpt->ProfileRemappingFile,
+                                        ThinOrFullLTOPhase::FullLTOPostLink));
```

Comment:
Don't think we need this one -- SampleProfileLoad during pre-link is only skipped in ThinLTO per https://github.com/llvm/llvm-project/blob/2e0cd6077a756fa28023d907045e6e53c50ec732/llvm/lib/Passes/PassBuilderPipelines.cpp#L1001-L1006 So for non-thin LTO it should have happened already.

Reply:
oh, you're right. This can basically just go back to the way it was then. Thanks for catching that.